File: Search Algorithms/Algorithms/BinarySearch.py
"""
  Performs binary search on an array
    with the given item and returns True or
    False.
    
    >>> search([5, 4, 1, 6, 2, 3, 9, 7], 2)
    True
    
    >>> search([5, 4, 1, 6, 2, 3, 9, 7], 8)
    False
"""

__author__ = "Kalpak Seal"

def binary_search(arr, x):
    arr = sorted(arr)

    first = 0
    last = len(arr) - 1
    found = False

    while first <= last and not found:
        mid = first + (last - first) // 2

        if arr[mid] == x:
            found = True
            break
        else:
            if x < arr[mid]:
                last = mid - 1
            elif x > arr[mid]:
                first = mid + 1

    return  found

if __name__ == "__main__":
    arr = [int(x) for x in input("Enter the integers of the list ").split()]
    x = int(input("Enter the number to be searched "))

    print(binary_search(arr, x))

File: Search Algorithms/Algorithms/BinarySearch.java
public class BinarySearch
{
    public int binarySearchIterative(int[] arr, int data)
    {
        int low = 0;
        int high = arr.length - 1;

        while(low <= high)
        {
            int mid = low + (high - low) / 2;
            if (arr[mid] == data)
                return mid;
                
            else if (arr[mid] < data)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return -1;
    }

    public int binarySearchRecursive(int[] arr, int low, int high, int data)
    {
        int mid = low + (high - low)/2;

        if(low > high)
            return -1;
        
        if (arr[mid] == data)
            return mid;
        
        else if (arr[mid] < data)
            return binarySearchRecursive(arr, mid + 1, high, data);
        else 
            return binarySearchRecursive(arr, low, mid - 1, data);
    }

    public static void main(String[] args) {
        int[] arr = {1,2,3,4,5,6,7,8,9};

        BinarySearch obj = new BinarySearch();

        System.out.println(obj.binarySearchIterative(arr, 8));
        System.out.println(obj.binarySearchRecursive(arr, 0, arr.length - 1, 2));
    }
}